diesel/src/pg/connection/mod.rs

```diff
@@ -109,35 +109,33 @@ impl Connection for PgConnection {
 impl PgConnection {
     #[cfg_attr(feature = "clippy", allow(type_complexity))]
     fn prepare_query<T: QueryFragment<Pg> + QueryId>(&self, source: &T)
-        -> QueryResult<(Rc<Query>, Vec<Option<Vec<u8>>>)>
+        -> QueryResult<(Rc<Statement>, Vec<Option<Vec<u8>>>)>
     {
         let mut bind_collector = RawBytesBindCollector::<Pg>::new();
         try!(source.collect_binds(&mut bind_collector));
         let binds = bind_collector.binds;
         let metadata = bind_collector.metadata;
 
-        let query = if source.is_safe_to_cache_prepared() {
-            let cache_len = self.statement_cache.len();
-            try!(self.statement_cache.cached_statement(source, &metadata, |sql| {
-                let query_name = format!("__diesel_stmt_{}", cache_len);
-                Query::prepare(
-                    &self.raw_connection,
-                    sql,
-                    &query_name,
-                    &metadata,
-                ).map(Rc::new)
-            }))
-        } else {
-            let mut query_builder = PgQueryBuilder::new();
-            try!(source.to_sql(&mut query_builder).map_err(Error::QueryBuilderError));
-            Rc::new(try!(Query::sql(&query_builder.finish(), Some(metadata))))
-        };
+        let cache_len = self.statement_cache.len();
```

Comment:
Why'd you move that out of the closure? Does borrowck not let you access this otherwise?